skrub/_datetime_encoder.py

```diff
@@ -18,62 +21,494 @@
+
+    Parameters
+    ----------
+    X_split : list of 1d array of length n_features
+        The 2d input, chunked into a list of array. This format allows us
+        to treat each column individually and preserve their dtype, because
+        dataframe.to_numpy() casts all columns to object when at least one
+        column dtype is object.
+
+    indices : list of int, default=None
+        Indices of the parsable columns to convert.
+        If None, indices are computed using the current input X.
+
+    indice_to_format : mapping of int to str, default=None
+        Dictionary mapping column indices to their datetime format.
+        It defines the format parameter for each column when calling
+        pd.to_datetime.
+
+        If indices is None, indices_to_format is computed using the current input X.
+        If format is not None, all values of indices_to_format are format.
+
+    format : str, default=None
+        Here for compatibility with ``pandas.to_datetime`` API.
+        When format is not None, it overwrites the values in indices_to_format.
+
+    Returns
+    -------
+    X_split : list of 1d array of length n_features
+    """
+    if indices is None:
+        indices, indice_to_format = _get_datetime_column_indices(X_split)
+
+    # format overwrite indices_to_format
+    if format is not None or indice_to_format is None:
+        indice_to_format = {col_idx: format for col_idx in indices}
+
+    for col_idx in indices:
+        X_split[col_idx] = pd.to_datetime(
+            X_split[col_idx], format=indice_to_format[col_idx], **kwargs
+        )
+
+    return X_split
+
+
+def _get_datetime_column_indices(X_split):
+    """Select the datetime parsable columns by their indices \
+    and return their datetime format.
+
+    Parameters
+    ----------
+    X_split : list of 1d array of length n_features
+
+    Returns
+    -------
+    datetime_indices : list of int
+        List of parsable column, identified by their indices.
+
+    indice_to_format: mapping of int to str
+        Dictionary mapping parsable column indices to their datetime format.
+    """
+    indices = []
+    indice_to_format = {}
+
+    for col_idx, X_col in enumerate(X_split):
+        X_col = X_col[pd.notnull(X_col)]
+        if _is_column_datetime_parsable(X_col):
+            indices.append(col_idx)
+            # TODO: pass require_dayfirst to _guess_datetime_format
+            indice_to_format[col_idx] = _guess_datetime_format(X_col)
+
+    return indices, indice_to_format
+
+
+def _is_column_datetime_parsable(X_col):
+    """Check whether a 1d array can be converted into a \
+    :class:`pandas.DatetimeIndex`.
+
+    Parameters
+    ----------
+    X_col : array-like of shape ``(n_samples,)``
+
+    Returns
+    -------
+    is_dt_parsable : bool
+    """
+    # Remove columns of int, float or bool casted as object.
+    try:
+        if np.array_equal(X_col, X_col.astype(np.float64)):
+            return False
+    except ValueError:
```

Comment:
this can raise a TypeError too
```python
import pandas as pd
from skrub import to_datetime

df = pd.DataFrame({"date": ["2023-10-12T17:36:50+00:00"]})
df["date"] = pd.to_datetime(df["date"])

to_datetime(df) # TypeError
```

Reply:
I guess it's now fixed, I can't reproduce it.